apps/arweave/src/ar_nonce_limiter.erl

```diff
@@ -1074,117 +1074,127 @@ apply_external_update2(Update, State) ->
+			CurrentSession2 = update_session(CurrentSession, StepNumber,
+					StepCheckpointsMap, Steps2),
+			State2 = apply_external_update4(State, SessionKey, CurrentSession2, Steps2),
+			{reply, ok, State2};
+		false ->
 			case IsPartial of
 				true ->
-					%% Inform the peer we have not initialized the corresponding session yet.
+					%% Inform the peer we miss some steps.
 					?LOG_DEBUG([{event, apply_external_vdf},
-						{result, session_not_found},
-						{vdf_server, ar_util:format_peer(Peer)},
-						{is_partial, IsPartial},
-						{session_key, encode_session_key(SessionKey)},
-						{server_step_number, StepNumber}]),
-					{reply, #nonce_limiter_update_response{ session_found = false }, State};
+							{result, missing_steps},
+							{vdf_server, ar_util:format_peer(Peer)},
+							{is_partial, IsPartial},
+							{session_key, encode_session_key(SessionKey)},
+							{client_step_number, CurrentStepNumber},
+							{server_step_number, StepNumber}]),
+					{reply, #nonce_limiter_update_response{
+							step_number = CurrentStepNumber }, State};
 				false ->
-					%% Handle the case where The VDF server has processed a block and re-allocated
-					%% steps from the previous session to the new session. In this case we only
-					%% want to apply new steps - steps in Session that weren't already applied as
-					%% part of PrevSession.
-
-					%% Start after the last step of the previous session
-					RangeStart = case get_session(PrevSessionKey, State) of
-						not_found -> 0;
-						PrevSession -> PrevSession#vdf_session.step_number + 1
-					end,
-					%% But start no later than the beginning of the session 2 after PrevSession.
-					%% This is because the steps in that session - which may have been previously
-					%% computed - have now been invalidated.
-					NextSessionStart = (SessionInterval + 1) * ?NONCE_LIMITER_RESET_FREQUENCY,
-					{_, Steps} = get_step_range(
-						Session, min(RangeStart, NextSessionStart), StepNumber),					
-					State2 = apply_external_update3(State, SessionKey, Session, Steps),
+					%% Handle the case where the VDF client has dropped of the
+					%% network briefly and the VDF server has advanced several
+					%% steps within the same session. In this case the client has
+					%% noticed the gap and requested the full VDF session be sent -
+					%% which may contain previously processed steps in a addition to
+					%% the missing ones.
+					%%
+					%% To avoid processing those steps twice, the client grabs
+					%% CurrentStepNumber (our most recently processed step number)
+					%% and ignores it and any lower steps found in Session.
+					{_, Steps} = get_step_range(Session, CurrentStepNumber + 1, StepNumber),
+					State2 = apply_external_update4(State, SessionKey, Session, Steps),
 					{reply, ok, State2}
-			end;
-		CurrentSession ->
-			#vdf_session{ step_number = CurrentStepNumber } = CurrentSession,
-			case CurrentStepNumber + 1 == StepNumber of
-				true ->
```

Comment:
It looks like we lost this clause in the refactoring. But I don't know if it's needed any more - I admit I can't full trace all the nested case statements to confirm the flow.

Reply:
Ah maybe this clause is just a special case of https://github.com/ArweaveTeam/arweave/pull/578/files#diff-f702581474afe210531b226e33de974a1134010509c618ed920cd12da463f82eR1151-R1156

Reply:
Yes, exactly